ash/src/entry.rs

```diff
@@ -140,15 +140,26 @@ impl Entry {
     /// `static_fn` must contain valid function pointers that comply with the semantics specified by
     /// Vulkan 1.0, which must remain valid for at least the lifetime of the returned [`Entry`].
     pub unsafe fn from_static_fn(static_fn: vk::StaticFn) -> Self {
-        let load_fn = |name: &std::ffi::CStr| {
+        let load_fn = move |name: &std::ffi::CStr| {
             mem::transmute((static_fn.get_instance_proc_addr)(
                 vk::Instance::null(),
                 name.as_ptr(),
             ))
         };
-        let entry_fn_1_0 = vk::EntryFnV1_0::load(load_fn);
-        let entry_fn_1_1 = vk::EntryFnV1_1::load(load_fn);
 
+        Self::from_parts_1_1(
+            static_fn,
+            vk::EntryFnV1_0::load(load_fn),
+            vk::EntryFnV1_1::load(load_fn),
+        )
+    }
+
+    #[inline]
+    pub unsafe fn from_parts_1_1(
```

Comment:
```suggestion
    pub fn from_parts_1_1(
```

Reply:
Mmm, `from_static_fn` is unsafe and this lets you do at least as much so why would it be safe?

Reply:
The same could be said for all other `from_parts_*` where we removed `unsafe`: these functions don't call anything unsafely, they just construct `Self`. Any other function on `Self` that invokes any of these functions is already marked `unsafe` for that reason.

Reply:
Ah I see now.

Reply:
> Any other function on Self that invokes any of these functions is already marked unsafe for that reason.

In `Entry`'s case that isn't actually true, which is why `from_static_fn` is unsafe. This should probably also be unsafe.

Reply:
Oh no! We also have `fn linked()` being safe, for obvious reasons, but then the following functions should have been marked `unsafe`:

- `try_enumerate_instance_version()`;
- `enumerate_instance_layer_properties()` (calls `entry_fn_1_0`);
- `enumerate_instance_extension_properties()` (^);
- `enumerate_instance_version()`.

I thought `from_static_fn()` would be `unsafe` because it calls a C function, not necessarily because it might "unsafely" initialize `Entry` hence not needing `unsafe` anymore on the functions listed above?

`create_instance()` and `get_instance_proc_addr()` are already `unsafe fn` too.

Reply:
With that I wouldn't trust the ICD/driver enough to not be `unsafe`, or is that why you made `fn linked()` safe?

So, all in all, what is your suggestion for the above? Make `Entry::from_parts_1_1()` `unsafe` again or make the methods on `Entry` unsafe?

Reply:
> With that I wouldn't trust the ICD/driver enough to not be `unsafe`

I'm not sure what "to not be `unsafe`" means here exactly. We must assume that they conform to the specifications, which means certain patterns of use are sound. When those patterns can be easily enforced statically, we can mark them as safe.

> is that why you made fn linked() safe?

`linked()` is safe because if you're replacing well known libraries with ones that behave incorrectly, you could just as well replace `libc` and break anything at all. The safety model doesn't extend arbitrarily ill-behaved environments.

> So, all in all, what is your suggestion for the above? Make Entry::from_parts_1_1() unsafe again or make the methods on Entry unsafe?

I'd be okay with either of these. I think I lean slightly towards the former, since it's a smaller change, the safe methods are pretty obviously safe, and you'd have to be doing something really dumb (supplying nonsensical function pointers) to make calling them unsafe. On the other hand, there's a case to be made that just making every wrapper unsafe is more consistent.

Reply:
> > With that I wouldn't trust the ICD/driver enough to not be `unsafe`
> 
> I'm not sure what "to not be `unsafe`" means here exactly. We must assume that they conform to the specifications, which means certain patterns of use are sound. When those patterns can be easily enforced statically, we can mark them as safe.

I hadn't expected Ash to validate that Vulkan's spec statements for these functions conform to Rust's notion of a "safe" function, and have thus far always been slapping `unsafe` on functions that are implemented outside of Rust and/or are nontrivial to validate the environment/arguments for.

> > is that why you made fn linked() safe?
> 
> `linked()` is safe because if you're replacing well known libraries with ones that behave incorrectly, you could just as well replace `libc` and break anything at all. The safety model doesn't extend to arbitrarily ill-behaved environments.

Okay.

> > So, all in all, what is your suggestion for the above? Make Entry::from_parts_1_1() unsafe again or make the methods on Entry unsafe?
> 
> I'd be okay with either of these. I think I lean slightly towards the former, since it's a smaller change, the safe methods are pretty obviously safe, and you'd have to be doing something really dumb (supplying nonsensical function pointers) to make calling them unsafe. On the other hand, there's a case to be made that just making every wrapper unsafe is more consistent.

Exactly, it feels odd to make the safe functions (in terms of implementation) `unsafe`, even though they allow other safe functions to behave `unsafe`ly when arbitrary pointers are stuffed into the function structs.

For consistency I'd also rather see the safe functions changed to `unsafe`. If we change `Entry::from_parts_1_1()` to be `unsafe` instead, we should do the same for `Instance` and `Device` in the event a safe function (that is influenced by such a constructor) is ever added.

Reply:
> and/or are nontrivial to validate the environment/arguments for.

Some functions are very straightforward! `unsafe` is the right default when there's any doubt, of course.

> it feels odd to make the safe functions (in terms of implementation) unsafe, even though they allow other safe functions to behave unsafely when arbitrary pointers are stuffed into the function structs.

This is a conventional use of `unsafe`. Safety contracts are easier to reason about when they don't span multiple items, but we pay that price either way here: either an API function is unsafe because maybe someone passed garbage to an `Entry` constructor in arbitrarily distant code, or the constructor is unsafe because maybe arbitrarily distant code will call a method after you pass garbage.

> For consistency I'd also rather see the safe functions changed to unsafe

Fine with me; consistency is important.

Reply:
> > and/or are nontrivial to validate the environment/arguments for.
> 
> Some functions are very straightforward! `unsafe` is the right default when there's any doubt, of course.

I don't think we can validate how straightforwardly they are implemented in a Vulkan driver? Maybe in the ICD though?

> This is a conventional use of `unsafe`. Safety contracts are easier to reason about when they don't span multiple items, but we pay that price either way here: either an API function is unsafe because maybe someone passed garbage to an `Entry` constructor in arbitrarily distant code, or the constructor is unsafe because maybe arbitrarily distant code will call a method after you pass garbage.

Exactly. But in this case I am concerned about the `unsafe`ness of calling that function even if it came from the driver/ICD. Otherwise we might have marked a whole bunch of functions (where we can trivially validate/restrict the argument range) safe.

> > For consistency I'd also rather see the safe functions changed to unsafe
> 
> Fine with me; consistency is important.

On the other hand it's `Entry` "inconsistently" marking some `extern "C"` (no matter how trivial) function wrappers as safe.

Reply:
> I don't think we can validate how straightforwardly they are implemented in a Vulkan driver?

The behavior dictated by the spec is the relevant concern. Many (most?) Vulkan functions specify complex soundness invariants, but some (like enumerating the instance version) are simple. Non-conformant drivers are intrinsically unsound just like a non-conformant libc would be.

> On the other hand it's Entry "inconsistently" marking some extern "C" (no matter how trivial) function wrappers as safe.

I'm not sure what you mean. I agree that marking methods on `Entry` as safe is appealing from a consistency standpoint, since ash as a matter of general policy does not attempt to reason about safety for you, though this doesn't have much to do with the presence of a FFI boundary per se.

Reply:
> I'm not sure what you mean. I agree that marking methods on `Entry` as unsafe is appealing from a consistency standpoint, since ash as a matter of general policy does not attempt to reason about safety for you, though this doesn't have much to do with the presence of a FFI boundary per se.

When observed from an FFI boundary, but you just said that that's irrelevant for `unsafe`ness when the soundness invariants of that function can be trivially upheld (and the function pointer is valid of course).

That would have otherwise been my reason to prefer keeping `from_parts` safe and marking the `"C"` functions unsafe, but I think you're right in that we should just mark `Entry::from_parts_1_1()` `unsafe` (explain why) and get this over with.

Reply:
Done now in https://github.com/ash-rs/ash/pull/756!